src/lib/hlarp.ml

```
@@ -175,34 +176,46 @@ module Athlates = struct
         else if found_header && line <> "" then
           let a1, a2 =
             Scanf.sscanf line "%s\t\t%s\t\t%f"
-              (fun allele1 allele2 conf ->
+              (fun allele1 allele2 confidence ->
                 { hla_class  = allele_to_hla_class allele1
                 ; allele     = allele1
-                ; qualifier = ""
-                ; confidence = 1.0 -. conf
+                ; qualifier  = ""
+                ; confidence
                 },
                 { hla_class  = allele_to_hla_class allele2
                 ; allele     = allele2
-                ; qualifier = ""
-                ; confidence = 1.0 -. conf   (* Don't REALLY understand their metric *)
+                ; qualifier  = ""
+                ; confidence
                 })
           in
-          let nacc =
-            match equal_pairs with
-            | `ReportAll -> a1 :: a2 :: acc
-            | `FirstPair -> if acc = [] then a1 :: a2 :: [] else acc
-            | `Unique    ->
-              let add_if_new a set = if List.mem a ~set then set else a :: set in
-              add_if_new a2 (add_if_new a1 acc)
-          in
-          loop true nacc
+          loop true (a1 :: a2 :: acc)
         else
           loop found_header acc
       with End_of_file ->
         close_in ic;
         acc
     in
     let alleles = loop false [] in
+    let alleles =
+      match equal_pairs with
+      | `ReportAll      -> alleles
+                        (* Will fail silently if < 2, I think ok. *)
+      | `FirstPair      -> List.take (List.rev alleles) 2
+      | `Unique         -> List.dedup alleles
```

Could this also fail silently?